Porting TF fake_quant_with_min_max functions

doncarlos999 opened the pull request:
Based on the discussion here: https://github.com/keras-team/keras/discussions/20319 I started porting the `fake_quant_with_min_max` functions from tensorflow to keras3.
This PR contains those ported functions and the relevant tests from https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/tests/fake_quant_ops_test.py.

I didn't implement `tf.quantization.fake_quant_with_min_max_vars` as it looks the same as `tf.quantization.fake_quant_with_min_max_args`. But, I can add this one too if required.

~For the CLA I am waiting on our CTO to add me to the Edge Impulse <-> Google CLA. But I figured that I can work on revisions to the PR in the meantime.~

CC: @matpalm, @dansitu, @james77777778

fchollet commented on keras/src/quantizers/quantizers.py:
```diff
@@ -127,6 +130,348 @@ def get_config(self):
         }
 
 
+@keras_export("keras.quantizers.adjust_and_nudge_quantization_range")
```
Could there be a shorter / friendlier name than `adjust_and_nudge_quantization_range`?

Also, does it need to be a public facing API?

fchollet commented on keras/src/quantizers/quantizers.py:
```diff
@@ -127,6 +130,348 @@ def get_config(self):
+    max_range: float = 6.0,
+    num_bits: int = 8,
+    narrow_range: bool = False,
+    name: Optional[str] = None,
```
Name argument is not used, you can remove it (same elsewhere)

fchollet commented on keras/src/quantizers/quantizers.py:
```diff
@@ -127,6 +130,348 @@ def get_config(self):
+@keras_export("keras.quantizers.fake_quant_with_min_max_args")
+def fake_quant_with_min_max_args(
+    inputs,
+    min_range: float = -6.0,
```
You don't really type annotations IMO -- we don't use those elsewhere

fchollet reviewed (commented):
Thanks for the PR!

james77777778 commented on keras/api/_tf_keras/keras/quantizers/__init__.py:
```diff
@@ -12,4 +12,14 @@
 from keras.src.quantizers.quantizers import compute_float8_scale
+from keras.src.quantizers.quantizers import fake_quant_with_min_max_args
+from keras.src.quantizers.quantizers import (
+    fake_quant_with_min_max_args_gradient,
```
For QAT purposes, I don't think we need `*_gradient` ops.

james77777778 commented on keras/api/_tf_keras/keras/quantizers/__init__.py:
```diff
@@ -12,4 +12,14 @@
+    fake_quant_with_min_max_vars_per_channel,
+)
+from keras.src.quantizers.quantizers import (
+    fake_quant_with_min_max_vars_per_channel_gradient,
```
For QAT purposes, I don't think we need `*_gradient` ops.

james77777778 commented on keras/api/quantizers/__init__.py:
```diff
@@ -12,4 +12,14 @@
 from keras.src.quantizers.quantizers import compute_float8_scale
+from keras.src.quantizers.quantizers import fake_quant_with_min_max_args
+from keras.src.quantizers.quantizers import (
+    fake_quant_with_min_max_args_gradient,
```
Same here.

james77777778 commented on keras/api/quantizers/__init__.py:
```diff
@@ -12,4 +12,14 @@
+    fake_quant_with_min_max_vars_per_channel,
+)
+from keras.src.quantizers.quantizers import (
+    fake_quant_with_min_max_vars_per_channel_gradient,
```
Same here.

james77777778 commented on keras/src/quantizers/__init__.py:
```diff
@@ -6,6 +6,16 @@
 from keras.src.quantizers.quantizers import compute_float8_scale
+from keras.src.quantizers.quantizers import fake_quant_with_min_max_args
+from keras.src.quantizers.quantizers import (
+    fake_quant_with_min_max_args_gradient,
```
Same here.

james77777778 commented on keras/src/quantizers/__init__.py:
```diff
@@ -6,6 +6,16 @@
+    fake_quant_with_min_max_vars_per_channel,
+)
+from keras.src.quantizers.quantizers import (
+    fake_quant_with_min_max_vars_per_channel_gradient,
```
Same here.

james77777778 commented on keras/src/quantizers/quantizers.py:
```diff
@@ -127,6 +127,341 @@ def get_config(self):
         }
 
 
+def adjust_and_nudge(min_range, max_range, num_bits, narrow_range):
```
Is `adjust_and_nudge` function following the implementation here?
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/fake_quant_ops_functor.h#L41
It seems slightly different.

Additionally, it might not work with `Model.fit` if the computation involves converting values to python `float` midway.

doncarlos999 replied:
Yes, I tried to match that implementation. 
I've removed that python `float` function and replaced it with `ops.cast`.

james77777778 commented on keras/src/quantizers/quantizers.py:
```diff
@@ -127,6 +127,341 @@ def get_config(self):
+    zero_point_from_min = quant_min_float - min_range / nudged_scale
+
+    # Determine nudged zero point
+    if zero_point_from_min < quant_min_float:
```
The if statement might not work in the XLA context. Could these lines be vectorized instead?

james77777778 commented on keras/src/quantizers/quantizers.py:
```diff
@@ -127,6 +127,341 @@ def get_config(self):
+    """Fake quantization operation matching TensorFlow's implementation."""
+
+    if isinstance(inputs, np.ndarray):
+        inputs = ops.convert_to_tensor(inputs)
```
It should be feasible to simply use `inputs = ops.convert_to_tensor(inputs)`

james77777778 commented on keras/src/quantizers/quantizers.py:
```diff
@@ -127,6 +127,341 @@ def get_config(self):
+
+
+@keras_export("keras.quantizers.fake_quant_with_min_max_args_gradient")
+def fake_quant_with_min_max_args_gradient(
```
For QAT purposes, I don't think we need `*_gradient` ops.

james77777778 commented on keras/src/quantizers/quantizers.py:
```diff
@@ -127,6 +127,341 @@ def get_config(self):
+@keras_export(
+    "keras.quantizers.fake_quant_with_min_max_vars_per_channel_gradient"
+)
+def fake_quant_with_min_max_vars_per_channel_gradient(
```
For QAT purposes, I don't think we need `*_gradient` ops.

james77777778 commented on keras/src/quantizers/quantizers_test.py:
```diff
@@ -100,3 +100,759 @@ def test_quantize_and_dequantize(self):
         # A loose assertion due to an expected quantization error
         self.assertAllClose(qdq_values, values, atol=5e-1)
+
+    def _TestOp(
```
We can use `@parameterized.named_parameters` and `named_product` to organize similar tests like this one:
https://github.com/keras-team/keras/blob/master/keras/src/ops/nn_test.py#L2355-L2365

doncarlos999 replied:
I have reduced the number of tests by merging the logic into a single function but if you still think it would help to use named_parameters to organize the tests then I can add it.

james77777778 replied:
I think we should follow the convention in the codebase by using `@parameterized.named_parameters` to organize the tests.

james77777778 commented on keras/src/quantizers/quantizers_test.py:
```diff
@@ -100,3 +100,759 @@ def test_quantize_and_dequantize(self):
+            num_bits=num_bits,
+            narrow_range=narrow_range,
+        )
+        self.assertAllClose(outputs, expected)
```
I think verifying the output values alone is not sufficient.
We can add an assertion for the gradient, similar to this one:
https://github.com/keras-team/keras/blob/master/keras/src/layers/core/dense_test.py#L584

james77777778 commented on keras/src/quantizers/quantizers_test.py:
```diff
@@ -100,3 +100,759 @@ def test_quantize_and_dequantize(self):
+        )
+        self.assertAllClose(outputs, expected)
+
+    def _TestGradOp(
```
For QAT purposes, I don't think we need `*_gradient` ops.

james77777778 commented on keras/src/quantizers/quantizers_test.py:
```diff
@@ -100,3 +100,759 @@ def test_quantize_and_dequantize(self):
+        )
+        self.assertAllClose(outputs, expected)
+
+    def _TestChannelsGradOp(
```
For QAT purposes, I don't think we need `*_gradient` ops.

james77777778 reviewed (commented):
Hi @doncarlos999 
I have left some comments.

Additionally, I think we still need `fake_quant_with_min_max_vars`, as it is used in TFMOT:
https://github.com/tensorflow/model-optimization/blob/master/tensorflow_model_optimization/python/core/quantization/keras/quant_ops.py#L340

james77777778 commented:
> Regarding the `*_gradient` functions, I added those as a way to test that the gradients that come from the main functions were being calculated correctly. Should we keep them just for testing purposes but not expose them in the public facing API? If not I will remove them.

We can test the gradients of `fake_*` functions using:
- tensorflow: `tf.GradientTape()` + `tape.gradient`
- torch: `loss.backward()` + `variable.grad`
- jax: `jax.grad`

You can refer to this test for an example:
https://github.com/keras-team/keras/blob/master/keras/src/layers/core/dense_test.py#L584-L649

Using a different function, separate from the user-facing function, for testing purposes seems redundant and fragile to me. However, we should wait for calls from @fchollet

doncarlos999 commented:
I agree that having two separate functions is fragile I simply kept the functions separate as that was how they were tested in the Tensorflow repo.
I will start adding tests based on your example in the meantime. Thank you.

doncarlos999 commented:
@james77777778 I have addressed your previous comments. 
I refactored the code into a single function to avoid duplication as the logic for the per_channel function on a 1 length array is the same as the other separate functions. If we need a separate function for handling parameters being passed in as float/int rather than single length arrays of floats/ints, then I can add the other function back (I have a previous commit that I can pull it back from).

fchollet commented:
Thanks for the updates!

@james77777778 should we merge?

james77777778 commented:
> Thanks for the updates!
> 
> @james77777778 should we merge?

Sorry for the late reply. I missed the updates.

Let me take a look...

james77777778 commented on keras/src/quantizers/quantizers.py:
```diff
@@ -127,6 +127,181 @@ def get_config(self):
+    #     return min_range, max_range, 1.0
+
+    # Calculate the step size
+    step_size = (max_range - min_range) / n_steps
```
We should use `step_size = ops.divide((max_range - min_range), n_steps)` to avoid issues when `max_range` or `min_range` is an int.

james77777778 commented on keras/src/quantizers/quantizers.py:
```diff
@@ -127,6 +127,181 @@ def get_config(self):
+        # Create gradient mask for all channels
+        masks = ops.cast(
+            (x >= qnt_min) & (x <= qnt_max),
+            dtype=np.float32,
```
should be `dtype="float32"`

james77777778 commented on keras/src/quantizers/quantizers.py:
```diff
@@ -127,6 +127,181 @@ def get_config(self):
+
+            # Gradient for min_val
+            # When x is clipped to min, the gradient flows to min_val
+            min_mask = ops.cast(x <= qnt_min, dtype=np.float32)
```
should be `dtype="float32"`

james77777778 commented on keras/src/quantizers/quantizers.py:
```diff
@@ -127,6 +127,181 @@ def get_config(self):
+
+            # Gradient for max_val
+            # When x is clipped to max, the gradient flows to max_val
+            max_mask = ops.cast(x >= qnt_max, dtype=np.float32)
```
should be `dtype="float32"`

james77777778 reviewed (commented):
@doncarlos999 
It’s great to merge three functions into one!

However, the implementation might be slightly different from TF’s because I had to set a higher tolerance to pass this simple test:

```python
import numpy as np
import tensorflow as tf

from keras.src.quantizers.quantizers import fake_quant_with_min_max_args
from keras.src.quantizers.quantizers import fake_quant_with_min_max_vars

# tf.quantization.fake_quant_with_min_max_args
x = np.random.uniform(low=-10, high=10, size=[10, 32, 32, 64]).astype("float32")
inp = tf.convert_to_tensor(x)
with tf.GradientTape() as tape:
    tape.watch(inp)
    tf_out = tf.quantization.fake_quant_with_min_max_args(
        inp, min=-6.0, max=6.0, num_bits=8
    )
tf_gradients = tape.gradient(tf_out, inp)

inp = tf.convert_to_tensor(x)
with tf.GradientTape() as tape:
    tape.watch(inp)
    keras_out = fake_quant_with_min_max_args(
        inp, min_vals=-6.0, max_vals=6.0, num_bits=8
    )
keras_gradients = tape.gradient(keras_out, inp)

np.testing.assert_allclose(tf_out.numpy(), keras_out.numpy(), atol=1e-5)
np.testing.assert_allclose(
    tf_gradients.numpy(), keras_gradients.numpy(), atol=1e-5
)

# tf.quantization.fake_quant_with_min_max_vars
x = np.random.uniform(low=-10, high=10, size=[10, 32, 32, 64]).astype("float32")
min = np.array(-5.5, dtype="float32")
max = np.array(7.5, dtype="float32")
tf_x = tf.convert_to_tensor(x)
tf_min = tf.Variable(min)
tf_max = tf.Variable(max)
with tf.GradientTape() as tape:
    tape.watch(tf_x)
    tape.watch(tf_min)
    tape.watch(tf_max)
    tf_out = tf.quantization.fake_quant_with_min_max_vars(
        tf_x, tf_min, tf_max, num_bits=8
    )
tf_gradients = tape.gradient(tf_out, [tf_x, tf_min, tf_max])

tf_x = tf.convert_to_tensor(x)
tf_min = tf.Variable(min)
tf_max = tf.Variable(max)
with tf.GradientTape() as tape:
    tape.watch(tf_x)
    tape.watch(tf_min)
    tape.watch(tf_max)
    keras_out = fake_quant_with_min_max_vars(tf_x, tf_min, tf_max, num_bits=8)
keras_gradients = tape.gradient(keras_out, [tf_x, tf_min, tf_max])

for tf_g, keras_g in zip(tf_gradients, keras_gradients):
    np.testing.assert_allclose(tf_g.numpy(), keras_g.numpy())

print("All passed!")

```

Could you please verify the impl?
Thanks!

doncarlos999 commented:
@james77777778 Thanks for the comments. I will work on revisions.

james77777778 commented on keras/src/quantizers/quantizers.py:
```diff
@@ -127,6 +127,144 @@ def get_config(self):
+    min_vals,
+    max_vals,
+    num_bits,
+    narrow_range,
```
We should add the default value `narrow_range=False` to be consistent with other APIs.

james77777778 commented on keras/src/quantizers/quantizers.py:
```diff
@@ -127,6 +127,144 @@ def get_config(self):
+):
+    """
+    Perform per-channel fake quantization with custom gradient using vectorized
+    operations.
```
To follow the docstrings convetion in Keras. We should use one line for this.
`Perform per-channel fake quantization.` should be enough. WDYT?

james77777778 commented on keras/src/quantizers/quantizers.py:
```diff
@@ -127,6 +127,144 @@ def get_config(self):
+
+
+@keras_export("keras.quantizers.fake_quant_with_min_max_vars_per_channel")
+def fake_quant_with_min_max_vars_per_channel(
```
I'm thinking that it might be possible to add a new argument `axis` to determine the channel axis.
This enhancement should improve user experience and seems feasible.

james77777778 replied:
This could be a following PR, I think

james77777778 reviewed (commented):
Hey @doncarlos999 Thanks for the updates!

I think this PR is almost ready to be merged.

However, I failed to pass this test using `tf.quantization.fake_quant_with_min_max_vars_per_channel`:

```python
import numpy as np
import tensorflow as tf

from keras.src.quantizers.quantizers import (
    fake_quant_with_min_max_vars_per_channel,
)


# tf.quantization.fake_quant_with_min_max_vars_per_channel
x = np.random.uniform(low=-10, high=10, size=[10, 32, 32, 3]).astype("float32")
min = np.array([-5.5] * 3, dtype="float32")
max = np.array([7.5] * 3, dtype="float32")
tf_x = tf.convert_to_tensor(x)
tf_min = tf.Variable(min)
tf_max = tf.Variable(max)
with tf.GradientTape() as tape:
    tape.watch(tf_x)
    tape.watch(tf_min)
    tape.watch(tf_max)
    tf_out = tf.quantization.fake_quant_with_min_max_vars_per_channel(
        tf_x, tf_min, tf_max, num_bits=8
    )
tf_gradients = tape.gradient(tf_out, [tf_x, tf_min, tf_max])

tf_x = tf.convert_to_tensor(x)
tf_min = tf.Variable(min)
tf_max = tf.Variable(max)
with tf.GradientTape() as tape:
    tape.watch(tf_x)
    tape.watch(tf_min)
    tape.watch(tf_max)
    keras_out = fake_quant_with_min_max_vars_per_channel(
        tf_x, tf_min, tf_max, num_bits=8, narrow_range=False
    )
keras_gradients = tape.gradient(keras_out, [tf_x, tf_min, tf_max])

for tf_g, keras_g in zip(tf_gradients, keras_gradients):
    np.testing.assert_allclose(tf_g.numpy(), keras_g.numpy())

print("All passed!")
``` 

Could you confirm the impl in `fake_quant_with_min_max_vars_per_channel`?
Thanks!

doncarlos999 commented:
@james77777778 Thanks for the feedback.
I went ahead and added the `axis` parameter in this PR as it helped fix the discrepancy with `tf.quantization.fake_quant_with_min_max_vars_per_channel`. I also addressed the other comments.

james77777778 reviewed (commented):
@doncarlos999 
On second thought, I think we only need to export one API since you’ve already combined the logic into a single impl.

Here’s my impl based on yours:

```python
...

from keras.src.backend.common.backend_utils import canonicalize_axis

...

@keras_export("keras.quantizers.fake_quant_with_min_max_vars")
def fake_quant_with_min_max_vars(
    inputs,
    min_vals,
    max_vals,
    num_bits,
    narrow_range=False,
    axis=None,
):
    """Perform per-tensor or per-channel fake quantization.

    `[min_vals, max_vals]` define the clamping range for the `inputs`.

    `inputs` are quantized into the quantization range (`[0, 2^num_bits - 1]`
    when `narrow_range=False` and `[1, 2^num_bits - 1]` when
    `narrow_range=True`) and then de-quantized and output as floats in
    `[min_vals, max_vals]` interval.

    This operation has a gradient and thus allows for training `min_vals` and
    `max_vals`.

    Args:
        inputs: Input tensor of float dtype.
        min_vals: A global minimum scalar or per-channel minimum tensor.
        max_vals: A global maximum scalar or per-channel maximum tensor.
        num_bits: Quantization bit width (e.g., `8` for int8).
        narrow_range: Whether to use narrow quantization range.
        axis: Axis along which to perform per-channel quantization. Set to
            `None` will perform per-tensor quantization. Defaults to `None`.

    Returns:
        Fake-quantized tensor
    """
    inputs = ops.convert_to_tensor(inputs)
    min_vals = ops.convert_to_tensor(min_vals)
    max_vals = ops.convert_to_tensor(max_vals)

    if axis is not None:
        axis = canonicalize_axis(axis, inputs.ndim)

    @ops.custom_gradient
    def _fake_quant_with_min_max_vars_per_channel(x, min_val, max_val):
        # Calculate quantization parameters for all channels at once
        nudged_min, nudged_max, scale, inv_scale = adjust_and_nudge(
            min_val, max_val, num_bits, narrow_range
        )

        quant_zero = ops.floor(
            ops.add(ops.multiply(-nudged_min, inv_scale), 0.5)
        )
        x_clamped = ops.clip(x, nudged_min, nudged_max)
        x_clamped_shifted = ops.subtract(x_clamped, nudged_min)
        result = ops.multiply(
            ops.floor(
                ops.add(
                    ops.subtract(
                        ops.multiply(x_clamped_shifted, inv_scale), quant_zero
                    ),
                    0.5,
                )
            ),
            scale,
        )

        # Create gradient mask for all channels
        masks = ops.cast(
            (x >= nudged_min) & (x <= nudged_max),
            dtype="float32",
        )

        def grad(*args, upstream=None):
            if upstream is None:
                (upstream,) = args

            # Gradient for x
            dx = ops.multiply(upstream, masks)
            axes = [i for i in range(len(dx.shape)) if i != axis]
            # Gradient for min_val
            # When x is clipped to min, the gradient flows to min_val
            min_mask = ops.cast(x <= nudged_min, dtype="float32")
            grad_min = ops.multiply(upstream, min_mask)
            if axis is not None:
                grad_min = ops.sum(grad_min, axis=axes)
            else:
                grad_min = ops.sum(grad_min)

            # Gradient for max_val
            # When x is clipped to max, the gradient flows to max_val
            max_mask = ops.cast(x >= nudged_max, dtype="float32")
            grad_max = ops.multiply(upstream, max_mask)
            if axis is not None:
                grad_max = ops.sum(grad_max, axis=axes)
            else:
                grad_max = ops.sum(grad_max)

            return dx, grad_min, grad_max

        return result, grad

    return _fake_quant_with_min_max_vars_per_channel(inputs, min_vals, max_vals)

```

The basic usage:

```python
import numpy as np
import tensorflow as tf

from keras.src.quantizers.quantizers import fake_quant_with_min_max_vars

# tf.quantization.fake_quant_with_min_max_vars
x = np.random.uniform(low=-10, high=10, size=[10, 32, 32, 64]).astype("float32")
min = np.array(-5.5, dtype="float32")
max = np.array(7.5, dtype="float32")
tf_x = tf.convert_to_tensor(x)
tf_min = tf.Variable(min)
tf_max = tf.Variable(max)
with tf.GradientTape() as tape:
    tape.watch(tf_x)
    tape.watch(tf_min)
    tape.watch(tf_max)
    tf_out = tf.quantization.fake_quant_with_min_max_vars(
        tf_x, tf_min, tf_max, num_bits=8
    )
tf_gradients = tape.gradient(tf_out, [tf_x, tf_min, tf_max])

tf_x = tf.convert_to_tensor(x)
tf_min = tf.Variable(min)
tf_max = tf.Variable(max)
with tf.GradientTape() as tape:
    tape.watch(tf_x)
    tape.watch(tf_min)
    tape.watch(tf_max)
    keras_out = fake_quant_with_min_max_vars(tf_x, tf_min, tf_max, num_bits=8)
keras_gradients = tape.gradient(keras_out, [tf_x, tf_min, tf_max])

for tf_g, keras_g in zip(tf_gradients, keras_gradients):
    np.testing.assert_allclose(tf_g.numpy(), keras_g.numpy())

# tf.quantization.fake_quant_with_min_max_vars_per_channel
x = np.random.uniform(low=-10, high=10, size=[10, 32, 32, 3]).astype("float32")
min = np.array([-5.5] * 3, dtype="float32")
max = np.array([7.5] * 3, dtype="float32")
tf_x = tf.convert_to_tensor(x)
tf_min = tf.Variable(min)
tf_max = tf.Variable(max)
with tf.GradientTape() as tape:
    tape.watch(tf_x)
    tape.watch(tf_min)
    tape.watch(tf_max)
    tf_out = tf.quantization.fake_quant_with_min_max_vars_per_channel(
        tf_x, tf_min, tf_max, num_bits=8
    )
tf_gradients = tape.gradient(tf_out, [tf_x, tf_min, tf_max])

tf_x = tf.convert_to_tensor(x)
tf_min = tf.Variable(min)
tf_max = tf.Variable(max)
with tf.GradientTape() as tape:
    tape.watch(tf_x)
    tape.watch(tf_min)
    tape.watch(tf_max)
    keras_out = fake_quant_with_min_max_vars(
        tf_x, tf_min, tf_max, num_bits=8, narrow_range=False, axis=-1
    )
keras_gradients = tape.gradient(keras_out, [tf_x, tf_min, tf_max])

for tf_g, keras_g in zip(tf_gradients, keras_gradients):
    np.testing.assert_allclose(tf_g.numpy(), keras_g.numpy())

print("All passed!")

```

WDYT?

doncarlos999 commented:
@james77777778 That looks good to me, do you want me to update the implementation? Or will you push those changes?

james77777778 commented:
> @james77777778 That looks good to me, do you want me to update the implementation? Or will you push those changes?

Feel free to update your impl if you think it looks good!

doncarlos999 commented:
@james77777778 Done.

doncarlos999 commented:
> Shouldn't fake_quant_with_min_max_args and fake_quant_with_min_max_vars_per_channel be removed?

Ah yes, my bad. Fixed

james77777778 commented:
LGTM!
Let's wait for @fchollet 's decision

fchollet reviewed (approved):
Thanks to both of you! LGTM

hertschuh commented:
Hi @doncarlos999 ,

It appears that these new tests are failing with JAX on GPU: https://source.cloud.google.com/results/invocations/a71c6dcb-3bf8-48ae-a712-d8410c3d64b8

Do you understand why?

james77777778 commented:
> Hi @doncarlos999 ,
> 
> It appears that these new tests are failing with JAX on GPU: https://source.cloud.google.com/results/invocations/a71c6dcb-3bf8-48ae-a712-d8410c3d64b8
> 
> Do you understand why?

Hi @hertschuh 
Please check out this PR: #20772
The root cause is that JAX on GPU has lower precision which caused the failure. The impact might be minor in the real world, so I skipped the assertions when using JAX GPU
